File: ios/Classes/TaxiLauncherPlugin.swift
import Flutter
import UIKit


private enum TaxiAppType: String {
    case uber
    case cabify
    case yango
    case careem
    case dtc
    case shail

    func type() -> String {
        return self.rawValue
    }
}

private class Taxi {
    let taxiAppName: String;
    let taxiAppType: TaxiAppType;
    let urlPrefix: String;


    init(taxiAppName: String, taxiAppType: TaxiAppType, urlPrefix: String) {
        self.taxiAppName = taxiAppName
        self.taxiAppType = taxiAppType
        self.urlPrefix = urlPrefix
    }

    func toMap() -> [String:String] {
        return [
            "taxiAppName": taxiAppName,
            "taxiAppType": taxiAppType.type(),
        ]
    }
}

private let taxis: [Taxi] = [
    Taxi(taxiAppName: "Uber", taxiAppType: TaxiAppType.uber, urlPrefix: "uber://"),
    Taxi(taxiAppName: "Careem", taxiAppType: TaxiAppType.careem, urlPrefix: "careem://"),
    Taxi(taxiAppName: "Cabify", taxiAppType: TaxiAppType.cabify,urlPrefix: "cabify://"),
    Taxi(taxiAppName: "DTC", taxiAppType: TaxiAppType.dtc,urlPrefix: "dtc://"),
    Taxi(taxiAppName: "Yango", taxiAppType: TaxiAppType.dtc,urlPrefix: "yango://"),
    Taxi(taxiAppName: "Shail", taxiAppType: TaxiAppType.shail,urlPrefix: "shail://")
]

private func getTaxiByRawTaxiAppType(type: String) -> Taxi? {
    return taxis.first(where: { $0.taxiAppType.type() == type })
}


private func showDirections(taxiType: TaxiAppType, url: String) {

        UIApplication.shared.open(URL(string:url)!, options: [:], completionHandler: nil)
}


private func isTaxiAppAvailable(taxi: Taxi?) -> Bool {
    // taxitype is not available on iOS
    guard let taxi = taxi else {
        return false
    }
    return UIApplication.shared.canOpenURL(URL(string:taxi.urlPrefix)!)
}


public class TaxiLauncherPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "taxi_launcher", binaryMessenger: registrar.messenger())
    let instance = TaxiLauncherPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }


    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getInstalledTaxis":
            result(taxis.filter({ isTaxiAppAvailable(taxi: $0) }).map({ $0.toMap() }))

        case "launchTaxi":
            let args = call.arguments as! NSDictionary
            let taxiType = args["taxiAppType"] as! String
            let url = args["url"] as! String

            let taxi = getTaxiByRawTaxiAppType(type: taxiType)
        
            if (!isTaxiAppAvailable(taxi: taxi)) {
                result(FlutterError(code: "TAXI_NOT_AVAILABLE", message: "Taxi app is not installed on a device", details: nil))
                return;
            }

            showDirections(
                taxiType: taxi!.taxiAppType,
                url: taxi!.urlPrefix + url
            )
            result(nil)

        case "isTaxiAppAvailable":
            let args = call.arguments as! NSDictionary
            let taxiType = args["taxiAppType"] as! String
            let taxi = getTaxiByRawTaxiAppType(type: taxiType)
            result(isTaxiAppAvailable(taxi: taxi))

        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

File: android/src/main/kotlin/com/example/taxi_launcher/TaxiLauncherPlugin.kt
package com.example.taxi_launcher

import android.content.Context
import android.content.Intent
import android.net.Uri
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


private enum class TaxiAppType { uber, careem, cabify, yango, dtc, yandexGo, grab, gojek, indrive }

private class TaxiAppModel(val taxiAppType: TaxiAppType, val taxiAppName: String, val packageName: String, val urlPrefix: String) {
    fun toMap(): Map<String, String> {
        return mapOf("taxiAppType" to taxiAppType.name, "taxiAppName" to taxiAppName, "urlPrefix" to urlPrefix)
    }
}

/** TaxiLauncherPlugin */
class TaxiLauncherPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "taxi_launcher")
        this.context = flutterPluginBinding.applicationContext
        channel?.setMethodCallHandler(this)
    }
    companion object {
        fun registerWith(
                @NonNull registrar: io.flutter.plugin.common.PluginRegistry.Registrar) {
            val taxiLauncherPlugin = TaxiLauncherPlugin()
            taxiLauncherPlugin.channel = MethodChannel(registrar.messenger(), "taxi_launcher")
            taxiLauncherPlugin.context = registrar.context()
            taxiLauncherPlugin.channel?.setMethodCallHandler(taxiLauncherPlugin)
        }
    }

    private val taxis = listOf(
            TaxiAppModel(TaxiAppType.uber, "Uber", "com.ubercab", "uber://"),
            TaxiAppModel(TaxiAppType.careem, "Careem", "com.careem.acma", "careeem://"),
            TaxiAppModel(TaxiAppType.cabify, "Cabify", "com.cabify.rider", "cabify://"),
            TaxiAppModel(TaxiAppType.yango, "Yango", "com.yandex.yango", "https://yango.go.link/en_int/order/"),
            TaxiAppModel(TaxiAppType.dtc, "DTC", "com.dtc.dtccustomer", "dtc://"),
            TaxiAppModel(TaxiAppType.yandexGo, "YandexGo", "com.yandex.go", "https://3.redirect.appmetrica.yandex.com/"),
            TaxiAppModel(TaxiAppType.grab, "Grab", "com.grabtaxi.passenger", "grab://"),
            TaxiAppModel(TaxiAppType.gojek, "Gojek", "com.gojek.app", "gojek://"),
            TaxiAppModel(TaxiAppType.indrive, "Indrive", "sinet.startup.inDriver", "indrive://")
    )

    private fun getInstalledTaxiApps(): List<TaxiAppModel> {
        return taxis.filter { taxi ->
            context.packageManager?.getLaunchIntentForPackage(taxi.packageName) != null
        }
    }

    private fun isTaxiAppAvailable(type: String): Boolean {
        val installedMaps = getInstalledTaxiApps()
        return installedMaps.any { taxi -> taxi.taxiAppType.name == type }
    }

    private fun launchTaxiApp(taxiAppType: TaxiAppType, url: String, result: Result) {
        context.let {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            val foundMap = taxis.find { map -> map.taxiAppType == taxiAppType }
            if (foundMap != null) {
                intent.setPackage(foundMap.packageName)
            }
            it.startActivity(intent)
        }
        result.success(null)
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "getInstalledTaxis" -> {
                val installedTaxies = getInstalledTaxiApps()
                result.success(installedTaxies.map { taxi -> taxi.toMap() })
            }

            "launchTaxi" -> {
                val args = call.arguments as Map<*, *>

                if (!isTaxiAppAvailable(args["taxiAppType"] as String)) {
                    result.error("TAXIAPP_NOT_AVAILABLE", "Taxi is not installed on a device", null)
                    return
                }

                val taxiAppType = TaxiAppType.valueOf(args["taxiAppType"] as String)
                val url = args["url"] as String

                launchTaxiApp(taxiAppType, url, result)
            }

            "isTaxiAppAvailable" -> {
                val args = call.arguments as Map<*, *>
                result.success(isTaxiAppAvailable(args["taxiAppType"] as String))
            }

            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}
